File: CPP/No690.cc
/**
 * Created by Xiaozhong on 2020/11/17.
 * Copyright (c) 2020/11/17 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>
#include <unordered_map>

using namespace std;

class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};

class Solution {
private:
    int dfs(int eid, unordered_map<int, Employee *> &mapper) {
        Employee *leader = mapper[eid];
        int ans = leader->importance;
        for (int subid : leader->subordinates) {
            Employee *e = mapper[subid];
            ans += dfs(e->id, mapper);
        }
        return ans;
    }

public:
    int getImportance(vector<Employee *> employees, int id) {
        if (employees.empty()) return 0;
        unordered_map<int, Employee *> mapper;
        for (Employee *employee : employees) mapper[employee->id] = employee;
        return dfs(id, mapper);
    }
};

File: Java/src/No690.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.*;

public class No690 {
    class Employee {
        public int id;
        public int importance;
        public List<Integer> subordinates;
    }

    public int getImportance(List<Employee> employees, int id) {
        Map<Integer, Employee> mapper = new HashMap<>();
        for (Employee employee : employees) mapper.put(employee.id, employee);
        Queue<Integer> queue = new ArrayDeque<>();
        int ans = 0;
        queue.offer(id);
        while (!queue.isEmpty()) {
            Employee employee = mapper.get(queue.poll());
            ans += employee.importance;
            for (int item : employee.subordinates) queue.offer(item);
        }
        return ans;
    }
}
